README.md
# NAME

PrimalNephia - Mini WAF

# SYNOPSIS

    ### Get started the PrimalNephia!
    $ primalnephia-setup MyApp

    ### And, plackup it!
    $ cd myapp
    $ plackup

# DESCRIPTION

PrimalNephia is a mini web-application framework.

__This software is still alpha quality. Its API may change without notice.__ 

# MOUNT A CONTROLLER

Use "path" function as following in lib/MyApp.pm .

First argument is path for mount a controller. This must be string.

Second argument is controller-logic. This must be code-reference.

In controller-logic, you may get Plack::Request object as first-argument,
and controller-logic must return response-value as hash-reference or PrimalNephia::Response object.

## Basic controller - Makes JSON response

Look this examples.

    path '/foobar' => sub {
        my ( $req ) = @_;
        # Yet another syntax is following.
        # my $req = req;

        return {
            name => 'MyApp',
            query => param('q'),
        };
    };

This controller outputs response-value as JSON, and will be mounted on "/foobar".

## Use templates - Render with Text::MicroTemplate

    path '/' => sub {
        return {
            template => 'index.html',
            title => 'Welcome to my homepage!',
        };
    };

Attention to "template" attribute.
If you specified it, controller searches template file from view-directory and render it.

If you use multibyte-string in response, please remember `use utf8` and, you may specify character-set as like as following.

    use utf8; ### <- very important
    path '/' => sub {
        return {
            template => 'mytemplate.tx',
            title => 'わたしのホォムペェジへよおこそ！',
            charset => 'Shift_JIS',
        };
    };

If you not specified `charset`, it will be 'UTF-8'.

## Makes response - Using "res" function

    path '/my-javascript' => sub {
        return res {
            content_type( 'text/javascript' );
            body( 'alert("Oreore!");' );
        };
    };

"res" function returns PrimalNephia::Response object with some DSL.

You may specify code-reference that's passed to res() returns some value. These values are passed into arrayref that is as plack response.

    path '/some/path' => sub {
        res { ( 200, ['text/html; charset=utf8'], ['Wooootheee!!!'] ) };
    };

And, you can write like following.

    path '/cond/sample' => sub {
        return res { 404 } unless param('q');
        return res { ( 200, [], ['you say '. param('q')] ) };
    };

Commands supported in "res" function are following.

- status
- headers
- header
- body
- content\_type
- content\_length
- content\_encoding
- redirect
- cookies

Please see Plack::Response's documentation for more detail.

## Limitation by request method - Using (get|post|put|del) function

    ### catch request that contains get-method
    get '/foo' => sub { ... };

    ### post-method is following too.
    post '/bar' => sub { ... };

    ### put-method and delete-method are too.
    put '/baz' => sub { ... };
    del '/hoge' => sub { ... };

## How to use routing with Router::Simple style matching-pattern and capture it - Using path\_param function

    post '/item/{id:[0-9]+}' => sub {
        my $item_id = path_param->{id}; # get param named "id" from path
        ...
    };

## Submapping to other applications that use PrimalNephia

It's easy. Call "path" function by package instead of a coderef.

    package MyApp;
    use PrimalNephia;

    path '/childapp' => 'ChildApp';

in MyApp/ChildApp.pm:

    package MyApp::ChildApp;
    use PrimalNephia;

    get '/message' => sub {
        message => 'this is child app!'
    };

This controller mapped to "/childapp/message".

Can use "+" prefix in package name. This syntax feature is to use absolute package name.

Example:

    package MyApp;

    path '/otherapp' => '+OtherApp';

"/otherapp" connect to "OtherApp".

Support to multiple path to SubApp.

Example:

    package MyApp;

    path '/subapp1' => 'SubApp';
    path '/subapp2' => 'SubApp';

The MyApp::SubApp connected to "/subapp1" and "/subapp2".

## Using Cookie

If you want to set cookie, use "set\_cookie" command.

    path '/somepath' => sub {
        set_cookie cookiename => 'cookievalue';
        return +{ ... };
    };

And, getting a cookie value, use "cookie" command.

    path '/anypath' => sub {
        my $cookie_value = cookie 'cookiename';
        return +{ ... };
    };

# USING CONFIG

First, see app.psgi that generated by `primalnephia-setup`.

    use strict;
    use warnings;
    use FindBin;

    use lib ("$FindBin::Bin/lib", "$FindBin::Bin/extlib/lib/perl5");
    use MyApp;
    MyApp->run;

You may define config with run method as like as following.

    MyApp->run(
      attr1   => 'value',
      logpath => '/path/to/log',
      ...
    );

And, you can access to these config in your application as following.

    path '/foo/bar' => sub {
      my $config = config;
    };

# STATIC CONTENTS ( like as image, javascript ... )

You can look static-files that is into root directory via HTTP.

# PLUGINS

You may use plugins for PrimalNephia.

See [PrimalNephia::Plugin](http://search.cpan.org/perldoc?PrimalNephia::Plugin) for more detail about plugin.

# FUNCTIONS

## path $path, $coderef\_as\_controller;

Mount controller on specified path.

## get post put del

Usage equal as path(). But these functions specifies limitation for HTTP request-method.

## req

Return Plack::Request object. You can call this function in code-reference that is argument of path().

## res $coderef

Return PrimalNephia::Response object with some DSL.

## param 

Return query-parameters that contains in path as hashref. (As like as "req->parameters")

## param $param\_name

Return specified query-parameter. (As like as "req->param($param\_name)")

## path\_param

Return parameters as hashref that captured by Router::Simple.

## path\_param $keyname

Return specified parameter that captured by Router::Simple.

## nip / nip $keyname

Alias for path\_param.

NOTE: Need more clever name for this function.

## config

Return config as hashref.

## base\_dir

Return the absolute path to root of application.

## cookie $cookie\_name

Get specified cookie value.

## set\_cookie $cookie\_name => $cookie\_value

Set value into specified cookie.

## nephia\_plugins @plugins

Load specified PrimalNephia plugins.

# AUTHOR

`ytnobody` <ytnobody@gmail.com>

`ichigotake`

`papix`

and PrimalNephia contributors, hachioji.pm

# SEE ALSO

[Plack::Request](http://search.cpan.org/perldoc?Plack::Request)

[Plack::Response](http://search.cpan.org/perldoc?Plack::Response)

[Plack::Builder](http://search.cpan.org/perldoc?Plack::Builder)

[Text::MicroTemplate](http://search.cpan.org/perldoc?Text::MicroTemplate)

[JSON](http://search.cpan.org/perldoc?JSON)

[PrimalNephia Home Page](http://nephia.github.io/)

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Nephia

{{$NEXT}}

0.37 2013-07-19T15:16:27Z
        - bug fix : fix documentation (could not use context() in load() of plugin)
        - feature : add plugin_config() DSL for development plugins

0.36 2013-07-19T14:14:44Z
        - refactor : modify some logic to more thin one
        - refactor : split plugin loader logic into Nephia::PluginLoader
        - refactor : add request object into context
        - refactor : enable context in before_action
        - feature : add before_action hook for plugins
        - refactor : suppress side-effect that destroys $env by Plack::Request
        - refactor : split action build logic into _build_action
        - bug fix : add Config::Micro as deps. "Keep genleness for user" :)
        - refactor : use trenary operator in _process_response

0.35 2013-07-17T11:53:56Z
        - feature : add application class name into context with keyname "app"
        - bug fix : fix strange specification of charset in http-header
        - refactor : use ternary operator in "path" feature
        - refactor : remove unused stuff

0.34 2013-07-17T01:58:21Z
        - refactor : move global-vars into Nephia::GlobalVars
        - feature  : add request context(Nephia::Context)
        - feature  : add hooks for development some plugins
        - feature  : add context feature for plugins
        - documentation : for context and hooks
        - documentation : split plugin documentation into Nephia::Plugin

0.33 2013-07-16T01:54:57Z
        - bug fix : suppress warning when mismatched overriden prototype [macopy]
        - bug fix : suppress a test "010_flavor.t" when OStype is windows

0.32 2013-07-12T02:15:43Z
        - bug fix : suppress warning when override DSL [ytnobody]
        - refactor : [INCOMPATIBLE] update refactor loading logic [songmu]

0.31 2013-07-11T04:06:33Z
        - bug fix : fix bug of reverse prefix in submapper [songmu]
        - bug fix : strange works of base_dir [songmu]
        - refactor : Implement Nephia::Response [songmu]
        - refactor : fix Nephia::Setup#get_version for more shortly [songmu]
        - refactor : remove duplicated code that loading plugins [songmu]
        - refactor : remove Nephia::ClassLoader and use Module::Load [songmu]
        - refactor : Nephia::Core inherits Exporter for export_to_level [songmu]
        - refactor : add test for Nephia::Core [songmu / ytnobody]
        - refactor : add template_path into arguments of Nephia::View when instantiate [ytnobody]
        - feature : --plugin option enabled in primalnephia-setup [papix]


0.30 2013-07-09T02:27:49Z
        - refactor : support multiple submapper[macopy]
        - bug fix : skip t/010_flavor.t when git is not installed

0.29 2013-07-08T06:29:54Z
        - bug fix : fix setup function [papix]

0.28 2013-07-08T05:58:20Z
        - refactor : modify setup more friendly for minilla
        - refactor : improved generation of cpanfile [papix]
        - bug fix : fix mojibake in config file [ichigotake]
        - feature : add Setup::Git as an only one core flavor [papix]

0.27 2013-07-06T17:07:26Z
        - refactor : add trigger "required_modules" in Nephia::Setup
        - feature : add nip / path_param feature
        - refactor : modify param function so that does not address the value associated with "Router::Simple".

0.26 2013-07-06T07:39:10Z
        - feature : modify to return the query parameter that specified when given an argument to param DSL 

0.25 2013-07-03T06:43:55Z
        - feature : add cookie / set_cookie DSL

0.24 2013-07-03T01:52:36Z
        - bug fix : fix logic that checks existance of plugin coderef

0.23 2013-07-02T03:21:31Z
        - refactor: Split DSLModifier as external module

0.22 2013-06-28T15:16:13Z
        - feature : add DSL Modifier

0.21 2013-06-28T05:46:17Z
        - feature : support ".primalnephia-setup" file in primalnephia-setup command
        - bug fix : fix base_dir test with psgi

0.20 2013-06-27T13:06:30Z
        - bug fix : sort flavor order by "view or not" rule

0.19 2013-06-27T07:13:20Z
        - feature : allow to use multiple flavor in primalnephia-setup

0.18 2013-06-26T16:07:12Z
        - bug fix : Nephia::MetaTemplate tag matching rule was not good :(

0.17 2013-06-26T15:04:27Z
        - bug fix : Nephia::MetaTemplate->argument added

0.16 2013-06-26T14:48:06Z
        - fix missing pseudo code for "res" in document
        - add meta template feature for creating new flavor easy

0.15 2013-06-21T02:33:36Z
        - add Config::Micro as a recommends
        - fix test for MSWin32

0.14 2013-06-19T07:34:12Z
        - bug fix : Setup::Base cpanfile requires Config::Micro

0.13 2013-06-19T07:13:42Z
        - add submapper feature [macopy]
        - wrap request-class with Nephia::request [songmu]
        - expand number of testfile
        - stop using 'Plack::Middleware::ContentLength' internally
        - implement get_version() feature into Setup.pm
        - show Nephia version to 'primalnephia-setup --version'

0.12 2013-06-14T11:11:42Z
        - import() will execute when plugin is loaded

0.11 2013-06-12T09:22:18Z
        - move Nephia to Nephia::Core for import() separation
        - add a feature that loads plugins when called import()
        - became able to use prototype-function in plugin

0.10 2013-06-10T12:12:12Z
        - Nephia::View::MicroTemplate is new default template engine. 
        - Nephia::View::Xslate was split to new distribution(it's same name).

0.09 2013-06-07T15:57:16Z
        - more quite design template and style
        - stop using Class::Load. use Nephia::ClassLoader
        - remove unrequired dependency (Mouse::Util::TypeConstraints, Class::Accesor::Fast)
        - stop using Path::Class.

0.08 2013-06-07T01:41:12Z
        - cut out validation feature into plugin (Nephia::Plugin::Data::Validator)
        - add documentation for commands in res function

0.07 2013-05-18T08:05:12Z
        - security fix for CVE-2013-1297

lib/PrimalNephia.pm
package PrimalNephia;
use strict;
use warnings;
use File::Spec;
our $VERSION = '0.37';

use PrimalNephia::Core;
use Module::Load ();

sub import {
    my ($class, %opts) = @_;
    my @plugins = ! $opts{plugins} ? () :
                  ref($opts{plugins}) eq 'ARRAY' ? @{$opts{plugins}} :
                  ( $opts{plugins} )
    ;

    PrimalNephia::Core->export_to_level(1);

    @_ = @plugins;
    goto do { PrimalNephia::Core->can('nephia_plugins') } if @plugins;
}

1;
__END__

=encoding utf8

=head1 NAME

PrimalNephia - Mini WAF

=head1 SYNOPSIS

  ### Get started the PrimalNephia!
  $ primalnephia-setup MyApp

  ### And, plackup it!
  $ cd myapp
  $ plackup

=head1 DESCRIPTION

PrimalNephia is a mini web-application framework.

B<This software is still alpha quality. Its API may change without notice.> 

=head1 MOUNT A CONTROLLER

Use "path" function as following in lib/MyApp.pm .

First argument is path for mount a controller. This must be string.

Second argument is controller-logic. This must be code-reference.

In controller-logic, you may get Plack::Request object as first-argument,
and controller-logic must return response-value as hash-reference or PrimalNephia::Response object.

=head2 Basic controller - Makes JSON response

Look this examples.

  path '/foobar' => sub {
      my ( $req ) = @_;
      # Yet another syntax is following.
      # my $req = req;

      return {
          name => 'MyApp',
          query => param('q'),
      };
  };

This controller outputs response-value as JSON, and will be mounted on "/foobar".

=head2 Use templates - Render with Text::MicroTemplate

  path '/' => sub {
      return {
          template => 'index.html',
          title => 'Welcome to my homepage!',
      };
  };

Attention to "template" attribute.
If you specified it, controller searches template file from view-directory and render it.

If you use multibyte-string in response, please remember C<use utf8> and, you may specify character-set as like as following.

  use utf8; ### <- very important
  path '/' => sub {
      return {
          template => 'mytemplate.tx',
          title => 'わたしのホォムペェジへよおこそ！',
          charset => 'Shift_JIS',
      };
  };

If you not specified C<charset>, it will be 'UTF-8'.

=head2 Makes response - Using "res" function

  path '/my-javascript' => sub {
      return res {
          content_type( 'text/javascript' );
          body( 'alert("Oreore!");' );
      };
  };

"res" function returns PrimalNephia::Response object with some DSL.

You may specify code-reference that's passed to res() returns some value. These values are passed into arrayref that is as plack response.

  path '/some/path' => sub {
      res { ( 200, ['text/html; charset=utf8'], ['Wooootheee!!!'] ) };
  };

And, you can write like following.

  path '/cond/sample' => sub {
      return res { 404 } unless param('q');
      return res { ( 200, [], ['you say '. param('q')] ) };
  };

Commands supported in "res" function are following.

=over 4

=item status

=item headers

=item header

=item body

=item content_type

=item content_length

=item content_encoding

=item redirect

=item cookies

=back

Please see Plack::Response's documentation for more detail.

=head2 Limitation by request method - Using (get|post|put|del) function

  ### catch request that contains get-method
  get '/foo' => sub { ... };

  ### post-method is following too.
  post '/bar' => sub { ... };

  ### put-method and delete-method are too.
  put '/baz' => sub { ... };
  del '/hoge' => sub { ... };

=head2 How to use routing with Router::Simple style matching-pattern and capture it - Using path_param function

  post '/item/{id:[0-9]+}' => sub {
      my $item_id = path_param->{id}; # get param named "id" from path
      ...
  };

=head2 Submapping to other applications that use PrimalNephia

It's easy. Call "path" function by package instead of a coderef.

  package MyApp;
  use PrimalNephia;

  path '/childapp' => 'ChildApp';

in MyApp/ChildApp.pm:

  package MyApp::ChildApp;
  use PrimalNephia;

  get '/message' => sub {
      message => 'this is child app!'
  };

This controller mapped to "/childapp/message".

Can use "+" prefix in package name. This syntax feature is to use absolute package name.

Example:

  package MyApp;

  path '/otherapp' => '+OtherApp';

"/otherapp" connect to "OtherApp".

Support to multiple path to SubApp.

Example:

  package MyApp;

  path '/subapp1' => 'SubApp';
  path '/subapp2' => 'SubApp';

The MyApp::SubApp connected to "/subapp1" and "/subapp2".

=head2 Using Cookie

If you want to set cookie, use "set_cookie" command.

  path '/somepath' => sub {
      set_cookie cookiename => 'cookievalue';
      return +{ ... };
  };

And, getting a cookie value, use "cookie" command.

  path '/anypath' => sub {
      my $cookie_value = cookie 'cookiename';
      return +{ ... };
  };

=head1 USING CONFIG

First, see app.psgi that generated by C<primalnephia-setup>.

  use strict;
  use warnings;
  use FindBin;

  use lib ("$FindBin::Bin/lib", "$FindBin::Bin/extlib/lib/perl5");
  use MyApp;
  MyApp->run;

You may define config with run method as like as following.

  MyApp->run(
    attr1   => 'value',
    logpath => '/path/to/log',
    ...
  );

And, you can access to these config in your application as following.

  path '/foo/bar' => sub {
    my $config = config;
  };

=head1 STATIC CONTENTS ( like as image, javascript ... )

You can look static-files that is into root directory via HTTP.

=head1 PLUGINS

You may use plugins for PrimalNephia.

See L<PrimalNephia::Plugin> for more detail about plugin.

=head1 FUNCTIONS

=head2 path $path, $coderef_as_controller;

Mount controller on specified path.

=head2 get post put del

Usage equal as path(). But these functions specifies limitation for HTTP request-method.

=head2 req

Return Plack::Request object. You can call this function in code-reference that is argument of path().

=head2 res $coderef

Return PrimalNephia::Response object with some DSL.

=head2 param 

Return query-parameters that contains in path as hashref. (As like as "req->parameters")

=head2 param $param_name

Return specified query-parameter. (As like as "req->param($param_name)")

=head2 path_param

Return parameters as hashref that captured by Router::Simple.

=head2 path_param $keyname

Return specified parameter that captured by Router::Simple.

=head2 nip / nip $keyname

Alias for path_param.

NOTE: Need more clever name for this function.

=head2 config

Return config as hashref.

=head2 base_dir

Return the absolute path to root of application.

=head2 cookie $cookie_name

Get specified cookie value.

=head2 set_cookie $cookie_name => $cookie_value

Set value into specified cookie.

=head2 nephia_plugins @plugins

Load specified PrimalNephia plugins.

=head1 AUTHOR

C<ytnobody> E<lt>ytnobody@gmail.comE<gt>

C<ichigotake>

C<papix>

and PrimalNephia contributors, hachioji.pm

=head1 SEE ALSO

L<Plack::Request>

L<Plack::Response>

L<Plack::Builder>

L<Text::MicroTemplate>

L<JSON>

L<PrimalNephia Home Page|http://nephia.github.io/>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
